### ISSUE TITLE
Case where highlighting doesn't terminate end of string in Atom, broken on Github

### ISSUE THREAD

**Issue by wesinator** [PowerShell/EditorSyntax#154]
### Environment

macOS
- Atom 1.34.0 (https://github.com/jrsconfitto/language-powershell)
- VS Code 1.30.2

### Issue Description

The included snippet of code highlights correctly in VS Code. 
It does not highlight at all on Github, and highlights incorrectly in atom. In Atom, the string highlighting is not terminated at the end of the string.

#### Screenshots

<img width="1369" alt="image" src="https://user-images.githubusercontent.com/5124946/51638679-063a8a80-1f2d-11e9-9313-445ab59702fa.png">

### Expected Behavior

Highlighted correctly on Github.
`),[IO.Compression.CompressionMode]::Decompress)),[Text.Encoding]::ASCII)).ReadToEnd()` highlighted as code in Atom.

### Code Samples

```PowerShell
sal a New-Object;iex(a IO.StreamReader((a IO.Compression.DeflateStream([IO.MemoryStream][Convert]::FromBase64String('aGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG8uIGhlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG8uICBoZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvLiAgIGhlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG8uICBoZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvCmhlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG8KCmhlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG8KCmhlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG8KaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvCgpoZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvCmhlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsb2hlbGxvaGVsbG9oZWxsbw=='),[IO.Compression.CompressionMode]::Decompress)),[Text.Encoding]::ASCII)).ReadToEnd()
```

**Comment by omniomi** [comment]
Linguist (used by Github's highlighting) is at the current revision for this project and so is VS Code so that's weird. 

I'll figure out what's going on with Atom but the Github issue may be related to Linguist specifically.

**Comment by msftrncs** [comment]
While VS Code shares Atom's textmate REGEX engine, I'd be willing to bet (but do not actually know) that Atom and Linguist share some common textmate portions, or at least share the same issue with handling textmate (maybe a line length buffer issue?), since they are both GitHub projects.  VS Code has a limit on how long of a line it will allow to be scoped, for performance reasons, but its like 20,000 characters.

@wesinator, Does it help to break up the string(since it has no affect on FromBase64String anyway)?

**Comment by wesinator** [comment]
Seems like a similar issue in Atom, using your example it works until you create a longer line length in the string

**Comment by wesinator** [comment]
May be related to https://github.com/atom/atom/issues/1667

**Comment by DA6IjY6jgHT8Z** [comment]
Is this related to the [atom-grammer-token-length](https://github.com/postcasio/atom-grammar-token-limit/issues/2) plugin? Does anybody know the fix for this plugin?

**Comment by msftrncs** [comment]
@DA6IjY6jgHT8Z, I don't think this limit can be exceeded.  Atom, and its Linquist engine have a line length limitation and that limitation appears to be strict at 1024 characters.  I can see how that might be of some use for a web application, but for practicality in a word wrapping code editor any limit is pointless.

The reason for limits on the processing of a line, or the fact that TextMate grammars only process one line at a time, is that regular expressions used to break down the language in to its parts could be poorly constructed and thus run very slowly, possibly consuming 10,000's of characters and hours of CPU time attempting to incorrectly break down one keyword or construct.

If I had a say on the matter I would use a processing steps constraint, such that if a single regular expression sequence exceeds a certain number of processing steps, then that step is aborted.  This process would need to be integrated directly in to the regex engine, since only it knows the number of steps its consumed.  I would then have no restriction to processing only 1 line at a time or even a line length limit.  I would also allow the grammar to be able to hint the appropriate limit when the need arises, and the application would still have an upper bound on the limit, since the app knows better external constraints (CPU resources) it has.

I personally feel that syntax highlighting is too important to just arbitrarily apply limits, and programming languages are too diverse to try to use reason for line length limits.   It is sad to see Atom's accepted solution to their self applied limit is to use a different grammar engine.

**Comment by DA6IjY6jgHT8Z** [comment]
Ok, thanks for clearing that up, I guess I'll just switch back to UltraEdit which has been doing this just fine for years. I am using this for web design and I need long paragraph support for content. I'm new to open source (and to programming in general) but was hoping there would be more flexibility. I may come back to this later when I have more training.
